src/token-provider.ts

```diff
@@ -1,46 +1,39 @@
+import { RetryStrategy } from './retry-strategy';
 export interface TokenProvider {
     fetchToken(): Promise<string>;
+    invalidateToken(token?: string);
 }
 
-export class SimpleTokenProvider implements TokenProvider {
-    constructor(public jwt: string) { }
-    fetchToken(): Promise<string> {
-        return new Promise<string>((resolve, reject) => {
-            resolve(this.jwt);
-        });
+/**
+ * Wrapper around the token provider that contains a retry strategy
+ * TODO: test
+ */
+export class RetryingTokenProvider implements TokenProvider {
+    constructor(
+        private baseTokenProvider: TokenProvider, 
+        private retryStrategy: RetryStrategy){}
+
+    fetchToken(){
+        return this.tryFetching();
     }
-}
 
-export function base64UrlDecode(encoded: string): string {
-    return atob(encoded.replace(/\-/g, '+').replace(/_/g, '/'));
-}
+    invalidateToken(token?: string){
+        this.baseTokenProvider.invalidateToken(token);
+    }
 
-export class AuthServerTokenProvider implements TokenProvider {
-    private accessToken: string = null;
-    constructor(private authServerUrl: string, private credentials?: string) { }
-    fetchToken(): Promise<string> {
-        return new Promise<string>((resolve, reject) => {
-            if (this.accessToken != null && Date.now() < JSON.parse(base64UrlDecode(this.accessToken.split(".")[1]))["exp"] * 1000) {
-                resolve(this.accessToken);
-            } else {
-                let xhr: XMLHttpRequest = new XMLHttpRequest();
-                xhr.onreadystatechange = () => {
-                    if (xhr.readyState === 4) {
-                        if (200 <= xhr.status && xhr.status < 300) {
-                            this.accessToken = JSON.parse(xhr.responseText)["access_token"];
-                            resolve(this.accessToken);
-                        } else {
-                            reject(new Error("Unexpected status code in response from auth server: " + xhr.status));
-                        }
-                    }
-                };
-                xhr.open("POST", this.authServerUrl, true);
-                xhr.setRequestHeader("Content-Type", "application/x-www-form-urlencoded");
-                xhr.send(
-                    "grant_type=client_credentials" +
-                    (this.credentials ? "&credentials=" + encodeURIComponent(this.credentials) : "")
-                );
-            }
+    private tryFetching(){
+        return new Promise<string>( (resolve, reject ) => {
+            this.baseTokenProvider.fetchToken()
+            .then( token => {
+                resolve(token);
+            }).catch( error => {
+                this.retryStrategy.attemptRetry(error, true)
+                .then(() => {
+                    this.tryFetching();
+                }).catch(error => { 
+                    reject(error); 
+            });
+        });
```

Comment:
Both `fetchToken` and `retryStrategy` use promises, couldn't we use them to simplify the retry interface?

```
this.retryStrategy.execute(() => {
  return this.baseTokenProvider.fetchToken()  
});
``` 

The callback would be executed on each retry and would return a promise. If that promise resolves, the result will be propagated by to the caller by the retry strategy. If there is an error, the retry strategy can decide whether to retry or propagate it. No need for "promise recursion" ;)